### PR TITLE
Loki: Fix error when empty template variables response

### LINKED ISSUE #69331: Dashboards: Loki template variables throw error when `label_values` returns nil result.
### 
# What went wrong?

**What happened**:

  - When loading a dashboard that has a template variable with a Loki datasource, and uses the `label_values` query, an error is displayed if the query does not return any values. (See screenshot)
image (8)

**What did you expect to happen**:

  - That template variables with Loki queries would behave the same as those with Prometheus queries. Namely that;
    - If the query returns no results, the template variable would remain empty and/or use the "all value" that is supplied.
    - The error would be caught and handled/discarded.

### 
# How do we reproduce it?

**Step 1**:

- Open Grafana create a new dashboard.

**Step 2**:

- Open the settings of the dashboard, and select "JSON Model"

**Step 3**:

- Add the following template variables under templating -> list, and click "Save changes" at the bottom.

```
{
        "current": {},
        "hide": 0,
        "includeAll": false,
        "multi": false,
        "name": "loki_datasource",
        "options": [],
        "query": "loki",
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "type": "datasource"
      },
      {
        "allValue": ".+",
        "current": {
          "text": [
            null
          ],
          "value": [
            null
          ]
        },
        "datasource": {
          "uid": "${loki_datasource}"
        },
        "definition": "",
        "error": {},
        "hide": 0,
        "includeAll": true,
        "label": "DAG file",
        "multi": true,
        "name": "dag_file",
        "options": [],
        "query": "label_values(dag_file)",
        "refresh": 2,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "tagValuesQuery": "",
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      }
```

**Step 4**:

- Note that the error in the screenshot is returned when the dashboard reloads.

### 
# What Grafana version are you using?

10.1.0-56420pre

### 
### Is the bug inside a Dashboard Panel?

Copy the panel's ["get-help" data](https://grafana.com/docs/grafana/latest/troubleshooting/send-panel-to-grafana-support/) here

### Grafana Platform?

I use Grafana Cloud

### User's OS?

Windows 11

### User's Browser?

Chrome

### Is this a Regression?

None

### Are Datasources involved?

Loki

### Anything else to add?

_No response_

### PR DESCRIPTION
**Pull request by ivanahuckova**
Fixes https://github.com/grafana/grafana/issues/69331

This PR fixes a bug introduced in https://github.com/grafana/grafana/pull/62109 where we changed `metadataRequest`
 to return `return res.data ?? (res || [])` instead of  `return res.data || [];`. This caused an error in cases, where responses didn't have any data and instead of returning `[]`, we returned `res` which caused issues, as we expected an array.

This was a reason why in https://github.com/grafana/grafana/issues/69331 we received an error, because we reseived response `{"status": "success"}` and we passed to it to processing of values in `labelValuesQuery`.

Therefore I have created specific method for `statsMetadataRequest`

Fixed: 

https://github.com/grafana/grafana/assets/30407135/94db5221-1c6c-4086-a814-c78a1767f19b

Broken:

https://github.com/grafana/grafana/assets/30407135/f5c26a88-8052-4f4b-942d-861709d0872d

### CODE REVIEW

**Review by matyax — COMMENTED** [review]
@ivanahuckova would you mind adding a regression test for this bug? That would also help to understand the code change. Thanks!

**Comment by ivanahuckova** [comment]
> @ivanahuckova would you mind adding a regression test for this bug? That would also help to understand the code change. Thanks!

I wanted to add a regression test, but I just can not come up with what the test should look like. The issue is that `\stats` endpoint returns non-typical response and therefore we need different metadata request. So the only thing I came up with is to test `getQueryStats` and check if `statsMetadataRequest` was called. But I don't think it adds any value as we mention that in the comment. So if you have any recommendation for test, let me know.

**Comment by matyax** [comment]
I would add a test case  for `statsMetadataRequest` and that should be enough.

**Review by matyax — APPROVED** [review]

**Comment by ivanahuckova** [comment]
Added simple test in https://github.com/grafana/grafana/pull/69373/files/0410d3f1a5ad02f2717cf2c334b33f728f967179..cb4acd0056ec8e1214184f42d650feb285224b7b

File: public/app/plugins/datasource/loki/datasource.test.ts [hunk 1]
```diff
@@ -1139,6 +1139,15 @@ describe('LokiDatasource', () => {
...
   });
+
+  describe('getQueryStats', () => {
+    it('uses statsMetadataRequest', async () => {
+      const ds = createLokiDatasource(templateSrvStub);
+      const spy = jest.spyOn(ds, 'statsMetadataRequest').mockResolvedValue({} as QueryStats);
+      ds.getQueryStats('{foo="bar"}');
+      expect(spy).toHaveBeenCalled();
+    });
+  });
```
Comment (matyax) [inline public/app/plugins/datasource/loki/datasource.test.ts:1150]: Sorry to be annoying. If you could add another case for expecting to throw it'd be ideal:

```
    if (url.startsWith('/')) {
      throw new Error(`invalid metadata request url: ${url}`);
    }
```

**Review by matyax — APPROVED** [review]
👏 👏

File: public/app/plugins/datasource/loki/datasource.test.ts (hunk 1 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/datasource.test.ts:1150]: Good idea! Added. :)

Comment (matyax) [reply public/app/plugins/datasource/loki/datasource.test.ts:1150]: Fantastic! Thanks 😄

**Pull request merged**